### PR TITLE
feat: add adjacency matrix in calorimeter clustering

### PR DESCRIPTION
**Pull request by wdconinc**
### Briefly, what does this PR introduce?
This introduces the adjacency matrix into calorimeter island clustering and is a port of https://github.com/eic/juggler/pull/2 into eicrecon. Copying the comments from original author @veprbl into this PR.

The issue with the current implementation is that `is_neighbour()` relies on geometrical dimensions of a calorimeter being used. This makes it cumbersome and error-prone to enable clustering for different detectors. The goal is to instead use readout cellID fields.

This PR suggests a following interface for doing that: A new parameter `adjacencyMatrix` is added to describe the adjacency of the sensitive elements. It can be used replace the existing settings like `localDistXY`. The value specified must be a string expression (to be evaluated by DD4hep evaluator) that returns a "boolean" value which is non-zero for adjacent cells. An example predicate for a calorimeter with a readout ID spec like `system:8,row:8,column:8` could look like

```python
cl = IslandCluster("cl",
  readoutClass="EcalBarrelSciGlassHits",
  adjacencyMatrix="(abs(row_1 - row_2) + abs(column_1 - column_2)) == 1",
  # ...
)
```

A calorimeter that has sectors might require a more sophisticated predicate like:
```
(
  abs(tower_1 - tower_2)
  + abs((sector_1 - sector2) * 5 + row_1 - row2)
) == 1
```

There are few minor issues with the current implementation:
1. The `5` is a magical constant for number of rows per sector. It is possible to scan the geometry and define some sort of `row_max` constant to be used in place.
2. Expressions are evaluated for every tower. This may be slow, but those values can be precomputed if needed.
3. The DD4hep evaluator is designed to work with doubles, there might be some floating points caveats. Should not affect us unless we get into large numbers?

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators @veprbl

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No.

### Does this PR change default behavior?
No.

### CODE REVIEW

**Comment by wdconinc** [comment]
Draft until CI succeeds (it wont).

File: src/detectors/BEMC/ProtoCluster_factory_EcalBarrelSciGlassProtoClusters.h [hunk 1]
```diff
@@ -33,6 +33,11 @@ class ProtoCluster_factory_EcalBarrelSciGlassProtoClusters : public JFactoryT<ed
         m_minClusterHitEdep=1.0 * dd4hep::MeV;    // from ATHENA reconstruction.py
         m_minClusterCenterEdep=30.0 * dd4hep::MeV; // from ATHENA reconstruction.py
 
+        // adjacency matrix
+        m_geoSvcName = "GeoSvc";
+        u_adjacencyMatrix = "";
```
Comment (wdconinc) [inline src/detectors/BEMC/ProtoCluster_factory_EcalBarrelSciGlassProtoClusters.h:38]: @veprbl Could you code-suggest what this should be and how I can see if it works here?

**Comment by wdconinc** [comment]
Marked as ready and asked for review since I think this is functional but have no ability to test without the settings that should be used for the sciglass.

File: src/detectors/BEMC/ProtoCluster_factory_EcalBarrelSciGlassProtoClusters.h (hunk 1 above)
Comment (veprbl) [reply src/detectors/BEMC/ProtoCluster_factory_EcalBarrelSciGlassProtoClusters.h:38]:
```suggestion
        u_adjacencyMatrix = "(abs(row_1 - row_2) + abs(column_1 - column_2)) == 1";
```

Comment (veprbl) [reply src/detectors/BEMC/ProtoCluster_factory_EcalBarrelSciGlassProtoClusters.h:38]: Oops, actually:
```suggestion
        u_adjacencyMatrix = "(abs(tower_1 - tower_2) + abs((sector_1 - sector2) * 5 + row_1 - row2)) == 1";
```

**Comment by wdconinc** [comment]
@veprbl I'm thinking we can avoid the cost by doing a cached lookup and only recalculate if not already calculated before. That's probably more relevant than a full pre-calculation. But we can accept the performance hit here at first and keep that improvement for a future pr.

**Comment by veprbl** [comment]
That is understood. Pre-calculation was also part of the idea that iterating over all of the geometry to figure out limits for the values.

**Comment by veprbl** [comment]
How do we enable island clustering in JANA2 now? In Gaudi there was a steering from python.

**Comment by wdconinc** [comment]
Isn't it already enabled? I thought it was. I'll look later, but I think this should be all that was necessary.

**Comment by wdconinc** [comment]
https://github.com/eic/EICrecon/blob/main/src/detectors/BEMC/BEMC.cc#L37 is where the factory is included which will then produce the collections of clusters using island clustering.

**Comment by wdconinc** [comment]
And https://github.com/eic/EICrecon/blob/main/src/services/io/podio/JEventProcessorPODIO.cc#L140 is where we tell Jana to require the sciglass clusters in the output, which means the factory will be asked to produce them.

**Comment by veprbl** [comment]
clustering

**Comment by wdconinc** [comment]
I just want to check if there were no regressions with calorimeter island clustering with the original method and then I'll merge.

**Comment by wdconinc** [comment]
No regressions in island clustering with other options:

### before
image

### after
image

**Review by veprbl — APPROVED** [review]

**Pull request merged**